src/engine/Server.cpp

```diff
@@ -775,18 +788,24 @@ Awaitable<T> Server::computeInNewThread(Function function) const {
 // _____________________________________________________________________________
 net::awaitable<Server::PlannedQuery> Server::parseAndPlan(
     const std::string& query, QueryExecutionContext& qec,
-    SharedCancellationHandle handle) const {
+    SharedCancellationHandle handle, TimeLimit timeLimit) const {
   return computeInNewThread([&query, &qec,
                              enablePatternTrick = enablePatternTrick_,
-                             handle = std::move(handle)]() {
+                             handle = std::move(handle), timeLimit]() {
     handle->resetWatchDogState();
     auto pq = SparqlParser::parseQuery(query);
     handle->throwIfCancelled("After parsing");
-    QueryPlanner qp(&qec);
+    QueryPlanner qp(&qec, handle);
     qp.setEnablePatternTrick(enablePatternTrick);
-    auto qet = qp.createExecutionTree(pq, *handle);
+    auto qet = qp.createExecutionTree(pq);
     handle->throwIfCancelled("After query planning");
-    return PlannedQuery{std::move(pq), std::move(qet)};
+    PlannedQuery plannedQuery{std::move(pq), std::move(qet)};
+
+    plannedQuery.queryExecutionTree_.getRootOperation()
+        ->recursivelySetCancellationHandle(std::move(handle));
+    plannedQuery.queryExecutionTree_.getRootOperation()
+        ->recursivelySetTimeConstraint(timeLimit);
```

Comment:
Is this the thing that we still need for the non-cancellable sorts?

Reply:
Yes